### PR TITLE
Tutorial: Snippet Sharing

### PR DESCRIPTION
**Pull request by Namoshek**
# Purpose

Tutorial that explains how to install, configure and run the Snippet Sharing tools. Task description in #1140.

# All Contents

- Scripts to configure, start and stop the rest-frontend
- Manpage for rest-frontend
- Mount specification of rest-backend config during build
- Readme improvements for both applications
- Tutorial

# Issues

- [ ] The mounted specification does not seem to work properly. The wildcard _ (underline) gets ignored by the rest-backend when querying keys with "current" instead of "_". Help welcome!

@markus2330 please review my pull request

### CODE REVIEW

**Comment by markus2330** [comment]
@tom-wa Can you help with the spec plugin (`_` seems to not work)?

@Namoshek I overlooked that there are still `_` in the key names. Can you use `/` instead? (e.g. `api/description`).

**Comment by Namoshek** [comment]
Thought so already when I saw the _ being used in spec. Will change it after lunch.

**Comment by Namoshek** [comment]
Do I have to create an extra mountpoint in order for spec to work properly, like it is done in most examples with `spec-mount` after the normal `kdb mount spec.ini /path ni`?

Btw. you can also ignore the spec and just import the following snippet:
```
$$> kdb import system/sw/elektra/restbackend/#0/current simpleini << EOF
backend/api/description/html = https://libelektra.org
backend/api/description/raw = https://raw.githubusercontent.com/ElektraInitiative/libelektra/master/doc/api_blueprints/snippet-sharing.apib
backend/jwt/encryption_key = some_random_string
backend/jwt/expiration_time = 7200
backend/kdb/path/configs = dir/configs
backend/kdb/path/users = dir/users
backend/output/default/entry/filterby = all
backend/output/default/entry/sort = asc
backend/output/default/entry/sortby = key
backend/output/default/user/filterby = all
backend/output/default/user/sort = asc
backend/output/default/user/sortby = username
backend/permissions/default_rank = 10
backend/permissions/entry/create = 10
backend/permissions/entry/delete = 50
backend/permissions/entry/edit = 50
backend/permissions/user/delete = 100
backend/permissions/user/edit = 100
backend/permissions/user/view = 100
cppcms/http/script_names/#0 = /
cppcms/service/api = http
cppcms/service/ip = 0.0.0.0
cppcms/service/port = 8080
EOF
```

**Comment by markus2330** [comment]
The spec-mount needs the metadata `mountpoint`, but you do not necessarily need an extra mountpoint within `spec`. I will do a review now, maybe I spot the mistake.

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+kdb set system/sw/elektra/restbackend/#0/current/backend/jwt/encryption_key "use_a_secret_key_here!"
+kdb set system/sw/elektra/restbackend/#0/current/backend/api_specification/html "http://link.to/the/html/version/of/api/description"
+kdb set system/sw/elektra/restbackend/#0/current/backend/api_specification/raw "http://link.to/the/blueprint/of/api/description"
+```
+
+In case you don't want to publish the API description, you can also set the keys
+to other links, e.g. your main page. It is good practice to have the description
+published though.
+
+To change the port of the server to 12345, you can use:
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: You first mentioned that some keys do not have defaults. You should make more clear what is the case here.

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+`kdb set system/sw/elektra/restbackend/#0/current/cppcms/service/port 12345`
+
+### Frontend ###
+
+The frontend does only require small mandatory changes, which have to be made in the
+`application-config.json` in the `/usr/local/share/elektra/tool_data/rest-frontend`
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: also mount it via yajl?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+To install npm via APT, use: `apt-get install npm`
+
+### Build the Applications ###
+
+After installing the dependencies, we are ready to build the applications.
+To do so, we can follow the steps explained in the [build guide](/doc/COMPILE.md).
+Make sure to include the two tools `rest-backend` and `rest-frontend`, e.g. by
+using the arguments `-DTOOLS=ALL` or `-DTOOLS=rest-backend;rest-frontend`.
+
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: We did not install it up to now?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+- Change `website.url` to the URL where the frontend will be reachable,
+  e.g. `http://libelektra.org/` (with trailing slash!)
+
+## Running the Applications ##
+
+As last step we need to run the applications:
+
+- First we start the backend server with `kdb run-rest-backend`.
+  To ensure the backend is accessible, you can use
+  `curl http://localhost:8080/version` (change port to your setting),
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: ``` `kdb get ...` ``` to get the correct setting?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  which should show you some version information in JSON format.
+- Although the frontend was compiled during installation already, we want to
+  have a freshly built homepage and use `kdb configure-rest-frontend` to do so.
+- Then we runt he frontend analogously with `kdb run-rest-frontend`. It should
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: runt?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  now be reachable at the configured port.
+
+If everything went smooth, both applications should now be online and reachable.
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: Seems to be really easy now, thanks!

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+Both applications can be stopped with a simple command:
+
+- Backend: `kdb stop-rest-backend`
+- Frontend: `kdb stop-rest-frontend`
+
+## Additional Tasks ##
+
+### API Specification ###
+
+The API of the backend was designed with an API blueprint very detailed.
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: sentence

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+- So we first install Ruby with gem: `apt-get install ruby`
+- Then we install the apiary-cli: `gem install apiaryio`
+- Finally we can use `apiary preview --path=snippet-sharing.apib --output=snippet-sharing-api.html`
+  in the [blueprints](/doc/api_blueprints) directory to generate the pretty version
+  of the API description. This file can then be placed wherever it is accessible
+  or needed.
+
+In case you change the API description (and the backend), you may want to ensure
+that your API blueprint is still syntax conform. To do so, you can use the tool
+[Drafter](https://github.com/apiaryio/drafter).
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: How to use drafter?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+### API Specification ###
+
+The API of the backend was designed with an API blueprint very detailed.
+To compile the [blueprint](/doc/api_blueprints/snippet-sharing.md) we need
+the [apiary-client](https://github.com/apiaryio/apiary-client) to be installed,
+which in return is installed as Ruby gem:
+
+- So we first install Ruby with gem: `apt-get install ruby`
+- Then we install the apiary-cli: `gem install apiaryio`
+- Finally we can use `apiary preview --path=snippet-sharing.apib --output=snippet-sharing-api.html`
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: Which directory?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+In case you change the API description (and the backend), you may want to ensure
+that your API blueprint is still syntax conform. To do so, you can use the tool
+[Drafter](https://github.com/apiaryio/drafter).
+
+### Use other Webserver than the built-in Grunt Webserver ###
+
+Of course it is possible to use another webserver instead of the built-in one.
+To do so, simply run `kdb configure-rest-frontend` and copy the content of the
+`public` directory in `/usr/local/share/elektra/tool_data/rest-frontend` to
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: `/usr/local/share/elektra/tool_data/rest-frontend/public` directory

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+It is preferred to install it along with [nodeJS](https://nodejs.org/).
+
+To install npm via APT, use: `apt-get install npm`
+
+### Build the Applications ###
+
+After installing the dependencies, we are ready to build the applications.
+To do so, we can follow the steps explained in the [build guide](/doc/COMPILE.md).
+Make sure to include the two tools `rest-backend` and `rest-frontend`, e.g. by
+using the arguments `-DTOOLS=ALL` or `-DTOOLS=rest-backend;rest-frontend`.
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: It does not get included via `-DTOOLS=ALL` for me? (Adding manually works)

File: src/tools/rest-backend/README.md [hunk 12]
```diff
@@ -118,40 +118,14 @@ should print information about the running service (PID, etc).
...
-- Download (latest) CppCMS from  [SourceForge](https://sourceforge.net/projects/cppcms/files/cppcms/)
-- Extract CppCMS: `tar -xjf cppcms-1.x.x.tar.bz2 && cd cppcms-1.x.x` (replace 1.x.x with your version)
-- Configure the build: `mkdir build && cd build && cmake ..`
-- Execute the build and run tests: `make && make test && make install`
-
-Further build and installation information can be found on their
-[website](http://cppcms.com/wikipp/en/page/cppcms_1x_build).
-It contains also a [guide](http://cppcms.com/wikipp/en/page/apt)
-explaining the installation through a dependency manager.
+An extensive [tutorial](/doc/tutorials/snippet-sharing-rest-service.md) is available for that.
```
Comment (markus2330) [inline src/tools/rest-backend/README.md]: for that?

File: doc/INSTALL.md [hunk 13]
```diff
@@ -66,7 +66,6 @@ Add the path where the library has been installed
 
 and run `ldconfig` as root.
 
-
```
Comment (Namoshek) [inline doc/INSTALL.md]: My first plan was to do so, but then I thought it seems out of place because no other plugin/tool/whatever is mentioned there (and I didn't notice I deleted one empty line accidentally until it was pushed, sorry).

Do you think a very general section like "Further Installation Manuals" would be fine?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+This tutorial explains how to install, configure and run everything that is required
+for the Snippet Sharing service and website. As operating system we expect Debian
+Jessie to be used, although most parts should also be applicable to other systems.
+
+## What is this? ##
+
+The Snippet Sharing service consists of two parts, a backend (the REST server) and
+a frontend (an AngularJS application executed in the browser), which does also house
+the Elektra website. The service itself can be used to share configuration snippets,
+i.e. configuration files of arbitrary applications.
```
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: In the tutorial or somewhere else for you to copy?

File: doc/INSTALL.md (hunk 13 above)
Comment (markus2330) [inline doc/INSTALL.md]: Yes, at the moment "Installation Manuals" sounds like a good idea, we can also add src/plugins/mozprefs/autoconfig/README.md there.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 11 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: I've not tested it so far, but what could be the reason for this behavior? Doesn't look intended to me.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 14 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: If it fits into the tutorial then you can add it there. Otherwise you can directly start with a new file in doc/news/. Btw. did you also see the news in doc/old?

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 3 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: Until now we prepared the installation by installing dependencies, yes. The installation itself is straight forward - include both tools and that's it.

Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: I mean you should mention it in the tutorial. A sentence and a link to doc/INSTALL.md should be enough.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 1 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: Good point, CppCMS requires manual configuration, almost forgot.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 2 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: Can I mount it even if it is not in the namespace specific configuration directory (e.g. `.dir`)?

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 4 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: We just set the port 10 lines above - I hope no one forgets it that fast. :smiley:

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 9 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: It is written one line below? ^^

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 8 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: I think that is out of scope for the tutorial and it is described quite well on their github page (therefore the link). When you have it installed, you simply run `drafter <file>`.

File: doc/INSTALL.md (hunk 13 above)
Comment (Namoshek) [inline doc/INSTALL.md]: That looks more like a configuration explanation, which doesn't seem to fit in the INSTALL.md?

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+use the command `kdb mount`. You should see a list of mountpoints, of which one should
+be something like `/sw/elektra/restbackend/#0`. If you do not see this mountpoint,
+use `kdb mount rest-backend-spec.ini /sw/elektra/restbackend/#0 ni` to mount it manually.
+
+After that you need to set three additional configuration parameters that have no defaults.
+It is recommended to set them for the system namespace if you will use a tool like
+`systemctl` to manage the services. (For the api_specification key, see below!)
+```
+kdb set system/sw/elektra/restbackend/#0/current/backend/jwt/encryption_key "use_a_secret_key_here!"
+kdb set system/sw/elektra/restbackend/#0/current/backend/api_specification/html "http://link.to/the/html/version/of/api/description"
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: In `doc/api_blueprints/snippet-sharing.apib` you proposed `restapi.libelektra.org` or is this only for the restapi itself? If so, we could use `doc.libelektra.org/restapi/$version` ($version as it is for the main api)?

Will there be any problems if we generate daniels and your parts together?

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 14 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: No, did not see them, but I just split the most recent news post as well (forgot before the merge). I can add the others as well.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 2 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: You could mount it into the dir-namespace, but then the user's workdirectory must be `/usr/local/share/elektra/tool_data/rest-frontend`, so it is better to mount it to system.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 4 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: You never know ;) We should at least use a consistent port.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 9 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: Above would be better!

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 3 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: I thought we are coming from the INSTALL.md, but I'll add one. ^^

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 8 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: Yes that would be enough, simply mention how to run it.

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+The applications configure themselves as much as possible during build,
+but some settings have to be set manually afterwards.
+
+### Backend ###
+
+To make sure that the configuration specification was set properly during installation,
+use the command `kdb mount`. You should see a list of mountpoints, of which one should
+be something like `/sw/elektra/restbackend/#0`. If you do not see this mountpoint,
+use `kdb mount rest-backend-spec.ini /sw/elektra/restbackend/#0 ni` to mount it manually.
```
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: Then I'll try this before making a ticket, maybe it is my fault.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 14 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: > I think I'll add a new news post.

Perfect! Do we also have to increment the version number or can we have a homepage release without Elektra release?

File: doc/INSTALL.md (hunk 13 above)
Comment (markus2330) [inline doc/INSTALL.md]: Yes, the "Installation Manuals"  section should be in doc/tutorials/README.md , but in INSTALL.md we could mention that some specific installations setups are described in doc/tutorials/README.md?

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 16 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: > maybe it is my fault.

For certain we could improve the usability, so opening such a ticket can never harm.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 2 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: That's not what I meant. If I use `kdb mount somefile.ini dir/ini simpleini` for example, it will create the file in the `.dir` folder. If I want to use a configuration file from an arbitrary folder, how do I do it?

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 14 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: Ok, then let us try to do the first non-version news!

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 2 above)
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: > If I want to use a configuration file from an arbitrary folder, how do I do it?

If you want it to a fixed file somewhere in the file system it would be `system`:
`kdb mount /usr/local/share/elektra/tool_data/rest-frontend system/sw/restfrontend/#0/current yajl`. A `kdb mount-restfrontend` would be nice.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 9 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: It actually reads better / proper if its not the raw markdown file (especially after I fixed the forced line-break in the enumeration).

File: doc/INSTALL.md (hunk 13 above)
Comment (Namoshek) [inline doc/INSTALL.md]: Sounds good!

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 15 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: `restapi.libelektra.org` was meant for the backend itself, not for additional resources. I think your suggestions with `doc.libelektra.org` sounds good.

I don't think you can "generate them together", but "putting the generated files in the same place" should be no issue at all.

File: doc/tutorials/snippet-sharing-rest-service.md (hunk 16 above)
Comment (Namoshek) [inline doc/tutorials/snippet-sharing-rest-service.md]: Described the issue in #1144.

**Comment by markus2330** [comment]
Ok, le uts get the tutorial finished. Please say when you are ready to merge. Further fixes please in separate PRs. (But please fix the conf. paths (w/o '_'), otherwise I'll have issues updating)

**Comment by Namoshek** [comment]
Also in rest-frontends application-config.json? I can't change
```
"translations": {
    "mappings": {
        "en_US": "en",
        "en_UK": "en"
    }
}
```
for example, because it obviously needs to have this names...

**Comment by markus2330** [comment]
> Also in rest-frontends application-config.json?

No, only in the parts you have control over. cppcms config itself could work (`/` -> `.`), too.

File: src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news.js [hunk 17]
```diff
@@ -57,7 +57,6 @@ module.exports = function (grunt) {
                     shortDesc = shortDesc[1];
                 } else {
                     shortDesc = '';
-                    grunt.log.warn('News post ' + post.name + ' has no description matching the configured regex');
```
Comment (markus2330) [inline src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news.js]: The warning seems to be useful? Could you please also document where logs are written to?

File: src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news-rss.js [hunk 18]
```diff
@@ -45,7 +45,7 @@ module.exports = function (grunt) {
             // filter guid from news post
             var guid = (new RegExp(self.data.regex.guid.pattern, self.data.regex.guid.flags)).exec(content);
             if (guid === null || guid.length < 2) {
-                grunt.log.error('Could not find `guid` in news, cannot create RSS file.');
```
Comment (markus2330) [inline src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news-rss.js]: Maybe a warning here, too? Just ignoring news that do not have a guid seems rude.

File: doc/tutorials/snippet-sharing-rest-service.md [hunk 19]
```diff
@@ -121,7 +121,7 @@ After that you need to set additional configuration parameters that have no defa
 It is recommended to set them for the system namespace if you will use a tool like
 `systemctl` to manage the services. (For the `api/description` keys, see below!)
 ```
-> kdb set system/sw/elektra/restbackend/#0/current/backend/jwt/encryption_key "use_a_secret_key_here!"
+> kdb set system/sw/elektra/restbackend/#0/current/backend/jwt/encryptionkey "use_a_secret_key_here!"
```
Comment (markus2330) [inline doc/tutorials/snippet-sharing-rest-service.md]: Maybe  `encryption/secret`? @petermax2 uses masterpassword, which you should also consider for consistency.

File: doc/news/2014-10-22_augeas.md [hunk 20]
```diff
@@ -0,0 +1,56 @@
+# Augeas and Config::Model #
```
Comment (markus2330) [inline doc/news/2014-10-22_augeas.md]: Seems like this is the first non-release post ;)

**Comment by markus2330** [comment]
Simply add "ready to merge" again if you feel like everything is ok.

File: src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news.js (hunk 17 above)
Comment (Namoshek) [inline src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news.js]: It is no log but basically stdout/stderr. I re-add the warnings... we should be able to get rid of this one if we add `- shortDesc: ` without content to news if they dont have a short description (which doesnt make sense, but ok).

File: src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news-rss.js (hunk 18 above)
Comment (Namoshek) [inline src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news-rss.js]: The reason are old news posts, but I quite quickly realized that most of them have not even a date, so I can't use them anyway.

File: doc/news/2014-10-22_augeas.md (hunk 20 above)
Comment (Namoshek) [inline doc/news/2014-10-22_augeas.md]: Yes, GSOC2014 would make a post too, but no date there.

File: src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news-rss.js (hunk 18 above)
Comment (markus2330) [inline src/tools/rest-frontend/resources/grunt-tasks/grunt-create-website-news-rss.js]: I would keep all (very) old things in one post.

**Comment by markus2330** [comment]
Thank you!
